cln_lspd.go

```diff
@@ -0,0 +1,281 @@
+	return nil, fmt.Errorf("not implemented")
+}
+
+func (s *server_c) CheckChannels(ctx context.Context, in *lspdrpc.Encrypted) (*lspdrpc.Encrypted, error) {
+
+	return nil, fmt.Errorf("not implemented")
+}
+
+//C-lightning plugin functions
+func StartPlugin() {
+	//c-lightning plugin initiate
+	plugin := glightning.NewPlugin(onInit)
+	plugin.RegisterHooks(&glightning.Hooks{
+		HtlcAccepted: OnHtlcAccepted,
+	})
+
+	err := plugin.Start(os.Stdin, os.Stdout)
+	if err != nil {
+		log.Printf("start plugin error: %v", err)
+	}
+}
+func onInit(plugin *glightning.Plugin, options map[string]glightning.Option, config *glightning.Config) {
+	log.Printf("successfully init'd! %v\n", config.RpcFile)
+
+	//lightning server
+	clientcln = glightning.NewLightning()
+	clientcln.StartUp(config.RpcFile, config.LightningDir)
+	log.Printf("successfull clientcln.StartUp")
+}
+
+func clnOpenChannel(clientcln *glightning.Lightning, paymentHash, destination string, incomingAmountMsat int64) (string, uint32, error) {
+
+	capacity := incomingAmountMsat/1000 + additionalChannelCapacity
+	if capacity == publicChannelAmount {
+		capacity++
+	}
+	minConf = 0 //need to be updated with mindepth for zeroconf
+
+	//open private channel
+	channelPoint, err := clientcln.FundChannelExt(destination, glightning.NewSat(int(capacity)), nil, false, &minConf, nil)
+	if err != nil {
+		log.Printf("clientcln.OpenChannelSync(%v, %v) error: %v", destination, capacity, err)
+		return "", 0, err
+	}
+	return channelPoint.FundingTxId, uint32(0), err
+}
+
+func clnIsConnected(clientcln *glightning.Lightning, destination string) error {
+	pubKey := destination
+	peers, err := clientcln.ListPeers()
+	if err != nil {
+		//log.Printf("clientcln.ListPeers() error: %v", err)
+		return fmt.Errorf("clientcln.ListPeers() error: %w", err)
+	}
+	for _, peer := range peers {
+		if pubKey == peer.Id {
+			log.Printf("destination online: %v", destination)
+			return nil
+		}
+	}
+	log.Printf("destination offline: %v", destination)
+	return fmt.Errorf("destination offline")
+}
+func OnHtlcAccepted(event *glightning.HtlcAcceptedEvent) (*glightning.HtlcAcceptedResponse, error) {
+	log.Printf("htlc_accepted called\n")
+	onion := event.Onion
+
+	log.Printf("htlc: %v\nchanID: %v\nincoming amount: %v\noutgoing amount: %v\nincoming expiry: %v\noutgoing expiry: %v\npaymentHash: %v\nonionBlob: %v\n\n",
+		event.Htlc,
+		onion.ShortChannelId,
+		event.Htlc.AmountMilliSatoshi, //with fees
+		onion.ForwardAmount,
+		event.Htlc.CltvExpiryRelative,
+		event.Htlc.CltvExpiry,
+		event.Htlc.PaymentHash,
+		onion,
+	)
+
+	paymentHash, paymentSecret, destination, incomingAmountMsat, outgoingAmountMsat, fundingTxID, fundingTxOutnum, err := paymentInfo([]byte(event.Htlc.PaymentHash))
+	if err != nil {
+		log.Printf("paymentInfo(%v)\nfundingTxOutnum: %v\n error: %v", event.Htlc.PaymentHash, fundingTxOutnum, err)
+	}
+	var zeroConfFundingTxID string
+	//var zeroConfFundingTxOutnum uint32
+	if hex.EncodeToString(paymentSecret) != "" {
```

Comment:
simpler to change the condition to paymentSecret != nil which eliminate the need to encode hex.
Here and in the following line.